src/server/builder.ts

```diff
@@ -0,0 +1,417 @@
+﻿/// <reference path="..\compiler\commandLineParser.ts" />
+/// <reference path="..\services\services.ts" />
+/// <reference path="protocol.d.ts" />
+/// <reference path="session.ts" />
+/// <reference types="node" />
+
+namespace ts.server {
+
+    interface Hash {
+        update(data: any, input_encoding?: string): Hash;
+        digest(encoding: string): any;
+    }
+
+    const crypto: {
+        createHash(algorithm: string): Hash
+    } = require("crypto");
+
+    /**
+     * An abstract file info that maintains a shape signature.
+     */
+    export class BuilderFileInfo {
+
+        private lastCheckedShapeSignature: string;
+
+        constructor(
+            public readonly scriptInfo: ScriptInfo,
+            public readonly project: Project
+        ) {
+        }
+
+        public isExternalModule() {
+            const sourceFile = this.getSourceFile();
+            return isExternalModule(sourceFile) || this.containsPurelyAmbientModules(sourceFile);
+        }
+
+        private containsPurelyAmbientModules(sourceFile: SourceFile) {

```

Comment:
please rename to `containsOnlyAmbientExternalModules` and add a comment why such files deserve special handling: they cannot be consumed from regular script files so there is no point to rebuild script files if such file is modified